src/goto-programs/goto_program.cpp

```diff
@@ -711,8 +713,26 @@ void goto_programt::instructiont::validate(
     DATA_CHECK(targets.empty(), "assign instruction should not have a target");
     break;
   case DECL:
+    DATA_CHECK_WITH_DIAGNOSTICS(
+      code.get_statement() == ID_decl,
+      "declaration instructions should contain a declaration statement",
+      source_location);
+    DATA_CHECK_WITH_DIAGNOSTICS(
+      !ns.lookup(to_code_decl(code).get_identifier(), table_symbol),
+      "declaring unknown symbol: " +
+        id2string(to_code_decl(code).get_identifier()),
+      source_location);
     break;
   case DEAD:
+    DATA_CHECK_WITH_DIAGNOSTICS(
+      code.get_statement() == ID_dead,
+      "dead instructions should contain a dead statement",
+      source_location);
+    DATA_CHECK_WITH_DIAGNOSTICS(
+      !ns.lookup(to_code_dead(code).get_identifier(), table_symbol),
+      "removing unknown symbol: " +
+        id2string(to_code_dead(code).get_identifier()) + " from scope",
```

Comment:
This part should go into the diagnostics bits instead of being part of the description (same applies to several other places).